warehouse/templates/includes/file-details.html

```diff
@@ -12,22 +12,42 @@
  # limitations under the License.
 -#}
 
-{% macro publisher(publ) -%}
+{% macro repo_url_with_reference(repo_url, reference) %}
+    {{ repo_url }}/tree/{{ reference }}
+{% endmacro %}
+
+{% macro workflow_filename(publisher) %}
+    {% if publisher.kind == "GitHub" %}
+        {{- publisher.workflow -}}
+    {% elif publisher.kind == "GitLab" %}
+        {{- publisher.workflow_filepath -}}
+    {% endif %}
+{% endmacro %}
+
+{% macro workflow_url(publisher, claims) %}
+    {%- set repo_url = claims.get('1.3.6.1.4.1.57264.1.12', '') -%}
+    {%- set build_hash = claims.get('1.3.6.1.4.1.57264.1.19', '') -%}
+    {%- set workflow_url = claims.get('1.3.6.1.4.1.57264.1.18', '') -%}
+    {%- set file_path = workflow_url.split('@')[0].replace(repo_url + '/', '') -%}
+    {{- repo_url + "/blob/" + build_hash + "/" + file_path -}}
+{% endmacro %}
+
+{% macro repository_name(publisher) %}
+    {{- publisher.repository -}}
+{% endmacro %}
+
+{% macro publisher(publ, claims) -%}
 {% if publ.kind == "GitHub" %}
-<p>
-  Publisher: <a href="https://github.com/{{ publ.repository }}/blob/HEAD/.github/workflows/{{ publ.workflow }}">
-    <i class="fa-brands fa-github" aria-hidden="true"></i>
-    <code>{{ publ.workflow }}</code> on {{ publ.repository }}
-  </a>
-</p>
+    {%- set brand="github" -%}
 {% elif publ.kind == "GitLab" %}
+    {%- set brand="gitlab" -%}
+{% endif %}
 <p>
-  Publisher: <a href="https://gitlab.com/{{ publ.repository }}/blob/HEAD/{{ publ.workflow_filepath }}">
-    <i class="fa-brands fa-gitlab" aria-hidden="true"></i>
-    <code>{{ publ.workflow_filepath }}</code> on {{ publ.repository }}
+  Permanent source link: <a href="{{ repo_url_with_reference(claims.get('1.3.6.1.4.1.57264.1.12'),claims.get('1.3.6.1.4.1.57264.1.13')) }}">
+    <i class="fa-brands fa-{{ brand }}" aria-hidden="true"></i>
+    {{ repository_name(publ) }}#{{ claims.get("1.3.6.1.4.1.57264.1.13") }}
```

Comment:
I think this is going to get confusing when we start to have build & publish attestations, since there's no guarantee that they come from the same hash. I think we should just push deep links like this down into the individual attestations instead.

Reply:
I've rolled back to the previous version, which only displayed the publisher and a path to the workflow file.

I liked having a prominent `permalink` (i.e. a link to the repository at the time of attestation) instead of a more generic one but it is still available in the details below.